add warning about print() and how to load layers in Processing algs

Gustry opened the pull request:
## Description

Just based on some discussions on the ML with @nyalldawson 

What should we do with https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/exampleprovider/ExampleAlgorithm.py#L36 ?
It's outdated, but it's still convenient for plugin writers. Can we maintain only one file with this one?

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

Gustry committed:
add hint about pushing information to the user in script template

nyalldawson commented on python/plugins/processing/script/ScriptTemplate.py:
```diff
@@ -149,6 +150,11 @@ def processAlgorithm(self, parameters, context, feedback):
 
+        # print(source.sourceCrs().authid())  # No
+        # Don't *ever* use "print()" from an algorithm.
+        # Use "feedback.pushInfo()" instead if you need to debug or tu push some information.
```
tu -> to

nyalldawson commented on python/plugins/processing/script/ScriptTemplate.py:
```diff
@@ -189,6 +195,17 @@ def processAlgorithm(self, parameters, context, feedback):
+        # If you really need to load some layers from this algorithm:
+        # QgsProject.instance().addMapLayers([source])  # No
+        # Don't *ever* do this from an algorithm, unless your algorithm returns
+        # the FlagNoThreading flag.It's not safe to manipulate project layers
```
space before "It's"

nyalldawson commented:
I'm unsure on this. I agree that this information is useful, but I'm reluctant to fill the example script up with too much dummy code which has to be removed. I think we may be better off to document all this nicely in the cookbook.

Gustry commented:
I fixed the typo for the `print` and removed the add layer after completion.
`print` is so widely used on the internet to debug a script, so I really think that it doesn't hurt to add this warning if you say that a print running in a thread is not safe.
If I'm spending a few minutes to debug a crashing script, a beginner in Python would not think about this simple 'print' statement.

Moreover, it shows the `pushinfo` which wasn't used yet in the template.

I keep the add layer to canvas in my notes for the cookbook indeed.

m-kuhn commented:
I think promoting good practice (as done with `pushInfo`) really is the most important thing.

What do you think about regexing a script on load for `print` statements and

* tell the user about the danger
* popping up a message to confirm and continue
* allow silencing the warning by adding a `# no-warn` to the print line

wonder-sk commented:
Just a random thought - what do you think about making the print() actually thread safe? It should not be difficult to defer resulting python console GUI updates to the main thread. We are starting to have more and more code executed in thread, so devs will probably hit this issue more often - so why not remove one bomb from the minefield called multi-threaded programming?

m-kuhn commented:
That will help to overcome the problems with print in threaded environments, but on the other hand will promote print as a safe option for debugging which it is not yet.

I think we

* either need to make print safe in every occasion (fix the python console to not call processEvents()), accepting that printing in loops is broken in 90% but not crashing in the other 10%
* or push devs towards using logging facilities rather than print

m-kuhn commented:
If #7344 is merged, I'm totally +1 for the proposal by @wonder-sk

Gustry commented:
Since #7344 is merged, I have again reduced my PR. I have added only the hint about feedback.pushinfo

m-kuhn commented:
+1 for merging as is.
Still, printing is not threadsafe and prints should be sent to the main loop's event queue (which can very well be done in a separate PR)

Gustry commented:
Hehe, the bot is missing: "check that committers are merging this PR if no more comments are coming" ;-)

Anything to fix? Or am I missing something?

m-kuhn commented:
No, good job, thanks everyone!
